### DISCUSSION TITLE
About the 8.0 version

### DISCUSSION THREAD

**Discussion by deguif** [ruflin/Elastica#1961] [Ideas]
Here's what I would like to propose:
- Create a new branch 7.x (branched from master)
- Create a new branch 8.x

For the 8.x branch, I propose to bump PHP version to 8.0 so that we can start using some recent PHP language evolutions (union type, named arguments, ...).

**Comment by ruflin** [comment]
I would make small adjustment to the proposal. Created a 7.x branch but no 8.x branch. 8.x is just master. Like this we need to do less backports.

**Reply by deguif** [reply]
Sounds good to me for the change ;)

**Reply by ruflin** [reply]
Do you want to create the branches?

**Reply by markuspoerschke** [reply]
Since, I do not see, that the branches are created yet, let me at some thoughts to it:

#### Proposal 1:

I suggest naming the branches as mentioned in the original post: `7.x` and `8.x`.

The advantages are:

1. By having semantic naming, developers can directly see, which version is developed in which branch.
2. Avoid [discussion](https://sfconservancy.org/news/2020/jun/23/gitbranchname/) about the `main` / `master` branch name.
3. The alias in the `composer.json` can then be removed, because Composer can detect the dev versions from the branch names.

#### Proposal 2:

Create the branches immediately.

That allows other developers to work on that new version 8. Otherwise, there is no branch to create pull requests against.

**Reply by deguif** [reply]
@markuspoerschke I like this approach, @ruflin @thePanz WDYT?
Many projects already use this approach about branches and versions.

**Reply by ruflin** [reply]
@markuspoerschke What do you do about `master / main` "Rename / fork" it and remove `master` afterwards?

**Reply by markuspoerschke** [reply]
I would say:

1. rename `master` to `7.x`
2. branch `8.x` from `7.x`

Which technically is the same as

1. branch `7.x` from `master`
2. branch `8.x` from `master`
3. delete `master` branch

It is of course possible to keep the existing `master` branch, but other than downloading development version, I do not see what can break if the `master` branch is deleted.

**Reply by ruflin** [reply]
I prefer to have as few branches as possible to remove confusion. Just learned that on Github there is a rename branch feature: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch I assume what this ensures is that open PR's also get the target branch adjusted. But this would imply to rather rename master to 8.x and branch 7.x from it.

That is also the only thing I don't like about the plan. Lets assume we get to 9.x and PR's are open. If I now branch 9.x from 8.x, the PR's will stay open against 8.x instead of 9.x. My personal preference is to only backport and ideally not forward port to prevent any PR's going into a previous version but don't get added to the "release" branch.

**Reply by deguif** [reply]
Or rename `master` to 7.x, and branch 8.x from 7.x as current open PRs target 7.x version

**Reply by markuspoerschke** [reply]
As a maintainer, the target branch of a PR can be changed. That should be the edge case. (I think the root of the problem is that there are stale PRs, like 80% of the PRs in this project are more than a year old.)

**Reply by ruflin** [reply]
You are right that the problem are the stale PR's. I'm glad the number of them is still pretty low.

On my end I'm very used to the main -> 7.x workflow but that is just old habit. Happy to try out the 8.x approach and see how it goes. As long as we just rename, always easy to go back.

**Reply by deguif** [reply]
@ruflin I can create the `8.x` branch, but I'm afraid I can't rename the `master` branch to `7.x` as I don't see the edit button on it?

**Reply by ruflin** [reply]
Seems like I have this edit button. But as `master` is the default branch I rather rename it to 8.x and the branch 7.x from it. Does this also make sense?

**Reply by thePanz** [reply]
I would rename it to `7.x`, as the current PRs are for the 7.x releases
As the 8.x branch just started, it is Ok (IMHO) to have it as a new branch.. wdyt?

**Reply by deguif** [reply]
👍  with @thePanz 
I was thinking the same ;)
> Or rename master to 7.x, and branch 8.x from 7.x as current open PRs target 7.x version

**Reply by ruflin** [reply]
The problem I have with this is that now we have to start forward port PRs. If we forget one, we have it in 7.x but no 8.x. Only 8.x is fine, only 7.x but not 8.x seems like an issue.

As the current PR's target master my thinking is these target 8.x. If we fork off 7.x from it, it will be in 7.x too.

**Comment by thePanz** [comment]
Should we define a list of "goals" for the new 8.0 version?
Maybe having a set of issues grouped into a milestone? WDYT?

My short list:
- More strict methods signatures (getting rid of arguments with "string|int|array|Class|null" as type)
- throwing exceptions when a call to ES failed (in some cases we return the Response, which contains errors)
- more?

**Reply by ruflin** [reply]
* Removal of deprecated code

We can use a milestone or just a meta github issue. Both works for me.

**Reply by damienalexandre** [reply]
As far as the wishlist goes:

- Clarify the usage / need of `elasticsearch/elasticsearch`: Elastica ship a Client, `elasticsearch/elasticsearch` too, and version 8 is even installing HttpPlug. There is also differences in the logging methods. Removing `elasticsearch/elasticsearch` could be nice IMO because I don't see why it's required by Elastica;

About the PHP requirement, I'm for keeping a maximum of compatibility, so 7.4 as minimum for now.

**Reply by ruflin** [reply]
@damienalexandre I added `elasticsearch/elasticsearch` as a dependency quite some time ago. The idea was to have it as an escape hatch if something is not supported by Elastica directly.

**Reply by coreation** [reply]
hey @ruflin just got see this 8.0 thread, does this mean if someone wants to use an 8.2 ElasticSearch feature, and there is compat. issue between Elastica 8.0 and ElasticSearch 8.2, we can just use the underlying dependency (elasticsearch/elasticsearch) to use that 8.2 feature?

**Reply by ruflin** [reply]
Potentially. The devil is always in the detail but likely yes. What you can always do is send "raw" json queries with Elastica.

**Comment by deguif** [comment]
- properties type-hinting
- get rid of custom encode/decode json handling (now that `JSON_THROW_ON_ERROR` will be available) 
- mark some classes as final (to simplify BC)
- review properties/method naming (removing `_` prefix for example)

About the version of PHP, I would require 8.0 as minimum version as 7.4 is in security support only maintenace mode. Many libraries already switched to 8.0 as minimum version. In the meantime we could maintain more regurlaly the 7.x branch for the ones that would be sticked to PHP 7.4 version or previous.

Advantage brought by PHP 8.0 would be for this library:
- union type
- `static` return type-hint (to avoid `@return this` annotation)
- named arguments (`json_decode` for example , it would avoid having to specify `$depth` when specifying custom flags)
- `mixed` type-hint
- `Stringable` interface
- ...

**Comment by ruflin** [comment]
Only supporting PHP 8.0 means that only users of PHP 8.0 can migrate to Elasticsearch 8.0. It is possible that Elastica 7.x will mostly work with Elasticsearch 8.0 but it is not guaranteed as there can be breaking changes.

One other data point is likely on what @ezimuel does with the official client on PHP version compatiblity?

**Reply by ezimuel** [reply]
We really would like to use the new features of PHP 8 for `elasticsearch-php` 8.0 but we still need to support PHP 7.4 until 28 Nov 2022. We have [many users](https://packagist.org/packages/elasticsearch/elasticsearch/php-stats) using PHP 7.4
That said, `elasticsearch-php`8.0.0 will still support PHP 7.4+. Btw, the `named argument` will be supported anymore since it's about how to call a function rather than how to write it. Moreover, we will finally use the class property type definition used by PHP 7.4.

**Comment by damienalexandre** [comment]
About Elasticsearch 8.0 compatibility, there is a new compatibility layer in Elasticsearch we could leverage while working on the next Elastica version: https://www.elastic.co/guide/en/elasticsearch/reference/8.0/rest-api-compatibility.html

By adding:

```
Accept: "application/vnd.elasticsearch+json;compatible-with=7"
Content-Type: "application/vnd.elasticsearch+json;compatible-with=7"
```

to all requests, the version 7 of Elastica could work with Elasticsearch 7 **and** 8.

**Reply by ruflin** [reply]
I'm on board adding these headers as it will ease the migration from 7 to 8. In general I hope in most cases, these are not even needed. 

Should we make this a config option in the connection or have it on by default?

**Reply by damienalexandre** [reply]
- It does not break anything if always there.
- The official client has an environment variable to enable it: https://github.com/elastic/elasticsearch-php/pull/1142

Both ways looks good to me.

**Reply by ruflin** [reply]
I like the idea of the environment variable and also it is in line with the official client.

**Reply by ezimuel** [reply]
@ruflin actually I need to add the API Compatibility header in elasticsearch-php v8. I rewrote the client from scratch supporting PSR-7 and PSR-8 standards and I need to re-apply some pieces. I opened this PR https://github.com/elastic/elasticsearch-php/issues/1221 for taking care of this.

**Reply by ruflin** [reply]
@ezimuel Why do you still need the header in 8? Is 8 compatible with 7?

**Reply by ezimuel** [reply]
@ruflin I think this is because we need to keep track of the version when we will release Elasticsearch 9.

**Comment by AndreasA** [comment]
Just wondering if there has been any news regarding 8.x support?

Also if you add compatibility support using enviornment variables, plase use `$_SERVER` and `$_ENV` instead of `getenv` to support symfony/dotenv and be thread-safe.

**Reply by ruflin** [reply]
Thanks for the ping @AndreasA We should start to make a move on the 8.x support. I was out for a few weeks so I'll ping @deguif @thePanz to see if any changes happened during that time also around branching?

**Reply by sidz** [reply]
@ruflin do you need any help? I open to invest some time into this package and help to move forward to ES 8

**Reply by ruflin** [reply]
@sidz Thanks for your offer, it would be great if could help out here. In this thread there were quite a few discussions, some with conclusions, some with open questions. Ideally we close this conversations and come to a conclusion to drive finally get 8.0 out of the door. Could you help with that?

**Comment by sidz** [comment]
Hey @ruflin. 
Based on conversation https://github.com/ruflin/Elastica/discussions/1961?sort=new#discussioncomment-3893793 here are simple steps which need to be done before we move forward and start to do something.

0. Release new 7.3? version in case there is smth to release
1. Rename master to 7.x
2. Create 8.x from 7.x
3. Delete master branch

after that:
1. Drop support for php <8.0 (7.4 is end of life)
2. Update official elasticsearch lib and add compatibility layer for 7.4 (Also here: improve CI pipeline and run library against ES 8)
3. Moving forward :)

**Reply by ruflin** [reply]
Thanks for putting together the summary @sidz . Seems like I'm outvoted on master to 7.x vs 8.x https://github.com/ruflin/Elastica/discussions/1961#discussioncomment-3017889 Lets do it then. My main ask is that by default, all existing and new PR's are merged first into 8.x and then backported to 7.x. I want to ensure that we don't have any feature in 7.x that are missing in 8.x by accident.  

* @thePanz Do yo you want to take on 0? 
* @deguif Interestined in taking 1., 2.. I wonder if 3 is needed as we could rename the branch directly on Github: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch Like this 1 will solve 3.

**Reply by thePanz** [reply]
Release 7.3 is prepared: https://github.com/ruflin/Elastica/pull/2126 @ruflin I will tag it as soon as we merge it :)

**Reply by thePanz** [reply]
Release tagged! :heavy_check_mark:

**Reply by deguif** [reply]
@ruflin Branch renaming looks appropriate.

**Reply by deguif** [reply]
Let's sum it up now that 7.3 is released:
- Create a 7.x branch from master
- Rename master to 8.x

Does that sound right?

On my side, I can create the branch for 7.x but I don't see the button for renaming the branch (follows https://github.com/ruflin/Elastica/discussions/1961#discussioncomment-3016511 )

**Reply by ruflin** [reply]
I just renamed master to 8.x. I know this is against what few of you proposed but I strongly believe that the open PR's should target 8.x by default and not 7.x. Anyone having open PR's can retarget their PR's to the upcoming 7.x branch if 8.x is the wrong branch (instead of the other way around). As far as I can see that is the only difference.

@deguif Can you take care of creating the 7.x branch from 8.x?

**Reply by deguif** [reply]
@ruflin it's done

**Reply by ruflin** [reply]
Thanks everyone. Looks like we are on track for phase 2 from the plan @sidz lined out:

```
* Drop support for php <8.0 (7.4 is end of life)
* Update official elasticsearch lib and add compatibility layer for 7.4 (Also here: improve CI pipeline and run library against ES 8)
* Moving forward :)
```

**Reply by sidz** [reply]
I could pick up first

**Reply by ruflin** [reply]
Go for it!

**Reply by mussbach** [reply]
That sounds really promising. Any news on this? What are your current plans for Elasticsearch 8?

**Reply by ruflin** [reply]
I think we dropped the ball a bit on this but should really pick it up again. @sidz Trying to get up-to-speed on where we left off?

**Reply by VincentLanglet** [reply]
Hi @ruflin, it's not easy for me to see the current status of the 8.x version.

Is there an updated to do list somewhere ?

**Reply by sidz** [reply]
@ruflin We stuck on the second option from my list. elasticsearch/elasticsearch package is not updated yet. And it won't be so easy to do due to new http layer and new namespaces.
see https://github.com/elastic/elasticsearch-php#backward-incompatible-changes-boom

**Reply by damienalexandre** [reply]
I know it's not the main topic discussed here, but may I suggest the complete removal of the `elastic/elasticsearch-php` dependency? It has evolves in a complex and self sufficient client, and I don't think it's relevant to build Elastica on top of it anymore.

Some arguments:

- it's used in very few spots in Elastica: https://github.com/search?q=repo%3Aruflin%2FElastica+%22use+Elasticsearch%22&type=code
- it release cycle is not the same / it can break BC
- it ships it's own Http client, like Elastica, so when a request is done, we don't know from a developer perspective which code is going to send the final request
- it's not only Http, everything is duplicated: the logs, the cluster configuration, the connection pool...
- we don't know if compatibility with OpenSearch is going to be a thing in future versions of `elasticsearch-php` ([see this](https://github.com/elastic/elasticsearch-py/pull/1623)). Elastica could be agnostic.

Refs: https://github.com/ruflin/Elastica/discussions/1961#discussioncomment-2550938

**Reply by ruflin** [reply]
The reason I made Elastica depend on elasticsearch-php is to get rid of a lot of code and have a fall back for all the functionality Elastica does not support. That is only used in a few spots is more because we never got do fully adopt it. In my ideal scenario, Elastica becomes a more high level API only and underlying is all elasticsearch-php and we can get rid of all the duplication (that was the goal). But of course that is not where we are today.

Around OpenSearch, it is also a question that was asked in the past for Elastica. Even though OpenSearch is initially a fork of Elasticsearch, the projects start to diverge future wise and likely also API wise. With the limited resources we have, I rather focus on only one product to keep compatibility and functionality in sync, which we already struggle. (for full disclosure, I work for Elastic, but Elastica is a side project for me not owned by Elastic)

@sidz About the migration to 8: I wonder if decoupling from elasticsearch-php would make any of this easier?

**Reply by ruflin** [reply]
For awareness, we have now a PR from @oleg-andreyev ready for 8.0 https://github.com/ruflin/Elastica/pull/2123 Can anyone interested please have a look there?

**Reply by sidz** [reply]
> About the migration to 8: I wonder if decoupling from elasticsearch-php would make any of this easier?

@ruflin obviously in this case we will have to write all API endpoints on our end :) It's a big part of work IMO. 
I didn't check but is there any new API between 8.x versions? In this case we will have to follow the same versioning like official SDK does.

I'm also voting against opensearch support.

**Reply by ruflin** [reply]
> I didn't check but is there any new API between 8.x versions? In this case we will have to follow the same versioning like official SDK does.

Elasticsearch so far doesn't use API versioning. There are new APIs for new features, but that is something we always have to deal with.

@damienalexandre @deguif @thePanz I'm planning to merge https://github.com/ruflin/Elastica/pull/2123 from @oleg-andreyev today, please let me now as soon as possible if you don't think this is the right path to go forward to get to an 8.0 release.

**Reply by oleg-andreyev** [reply]
Things got merged by @ruflin recently, so we can continue with testing, we've added 8.x to CI and also testing against nightly-builds (dev)

**Reply by coreation** [reply]
Hey everyone, I was wondering how the 8.x branch was coming along and if there's a roadmap somewhere with "things to be done"?  If possible, I could see how I could help out to push an ES8 compatible release out.

**Reply by ruflin** [reply]
@coreation The 8.x branch is coming along nicely. As usual with every "major" release the goal is to get rid of all the deprecated bits to be able to simplify the code. But also, with each major release we fail to do this completely (which is ok). If you could go through the code base and check for deprecated bits and potentially open a PRs that cleans it up, this would be great.

One thing that I think is blocking is the current Elasticsearch version dependency we are using: https://github.com/ruflin/Elastica/blob/8.x/composer.json#L20 This should go to 8: https://github.com/elastic/elasticsearch-php I remember we had discussions around this as it contains breaking changes. Could you have a look there on what is needed? Likely just opening a PR with the change will already show what is breaking and what not.

**Reply by ruflin** [reply]
++ on getting out a 8.0 release rather soonish.

**Reply by coreation** [reply]
Hey @ruflin thanks for the pointers, I'll have some time over the weekend and dedicate some of it to have a look. To be more concrete, I'll check the [breaking changes](https://github.com/elastic/elasticsearch-php#backward-incompatible-changes-boom) that the elasticsearch-php library brings with their 8 version and see how it affects the 8.x branch.

**Reply by coreation** [reply]
@ruflin  I found some time to check for (small) deprecations, the PR for that is available [here](https://github.com/ruflin/Elastica/pull/2167).

**Reply by coreation** [reply]
@ruflin the PR (WIP) for the elasticsearch dependency upgrade can be found [here](https://github.com/ruflin/Elastica/pull/2168).

**Reply by coreation** [reply]
@ruflin I've added a question wrt to the refactor approach on the last mentioned PR...quite some decisions and work to be done for sure.

**Reply by ruflin** [reply]
Thanks for start tackling this. Will chime in on the PR to discuss further, there might be some tricky decisions we have to make.

**Reply by coreation** [reply]
Sounds good, from my - layman - point of view, there's indeed some foundational choices we need to make before we can start refactoring things.

**Reply by mussbach** [reply]
As the final Version of the 7 branch Version 7.17.x will only be supported until 2023-08-01 (https://www.elastic.co/support/eol). I see some good progress in https://github.com/ruflin/Elastica/pull/2168 but I do not have a feeling about what is still left. So let me please just ask whether you think that support for 8 will be implemented before 7 is eol?

**Reply by ruflin** [reply]
7.17.x is supported until "The later of 2023-08-01 or the release date of 9.0.0". As there is no 9.0 out yet, 7.17.x will stay supported is my understand.

@mussbach To your question, I strongly hope we make some quick progress on moving to 8.0 but at the same time it is quite a bit of effort. Can't give you an estimate.

**Reply by mussbach** [reply]
Sorry to constantly come back to this. Any news on the update to elasticsearch 8? Alternatively: Do you know any other lib that supports elasticsearch 8 for php?

**Reply by ruflin** [reply]
We are close on getting the major change to 8.x in: https://github.com/ruflin/Elastica/pull/2181 @mussbach Would you be willing to help testing / trying it out? Otherwise, there is the official elasticsearch php client which should work with 8.x.

**Reply by mussbach** [reply]
Nice, this is good news! I can definitely test and probably help to fix some bugs, if there are some suitable ones.

**Reply by ruflin** [reply]
@mussbach Great to hear. Please subscribe to the PR so you stay up-to-date.

**Comment by ruflin** [comment]
A major pull request to get us to version 8 compatibility has just been merged: https://github.com/ruflin/Elastica/pull/2181 Please help with testing, reporting bugs and bug fixes. Still a bit of work ahead of us tracked in https://github.com/ruflin/Elastica/issues/2185

**Comment by ruflin** [comment]
Release 8.0.0 is out: https://github.com/ruflin/Elastica/releases/tag/8.0.0 Going to close this thread. Thanks for everyone that helped us to get here. Next step is upgrading projects / libs and get all the feedback in.
